File: frontend_integration/precision_pricing.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */

/**
 * APEX Ω — Precision Pricing Engine
 *
 * Canonical responsibilities:
 *
 * 1. Resolve token decimals.
 * 2. Resolve oracle decimals.
 * 3. Normalize every oracle price into PRICE_SCALE.
 * 4. Convert token atomic units into USD fixed-point units.
 * 5. Convert USD values back into token atomic units.
 * 6. Derive TOKEN_A / TOKEN_B cross prices.
 * 7. Reject stale, invalid, incomplete, or divergent prices.
 * 8. Preserve integer precision throughout the execution pipeline.
 *
 * No JavaScript Number is used for monetary arithmetic.
 *
 * This is the reference implementation. The Python port in
 * omega_v5/pricing/precision_pricing.py is kept bit-for-bit equivalent
 * for pipeline compatibility.
 */

export type Address = `0x${string}`;

export const PRICE_DECIMALS = 18;
export const PRICE_SCALE = 10n ** BigInt(PRICE_DECIMALS);

export const BPS_DENOMINATOR = 10_000n;

export type OracleKind =
  | "CHAINLINK"
  | "DEX_TWAP"
  | "PROTOCOL_RATE"
  | "MANUAL_DISABLED";

export interface TokenMetadata {
  chainId: number;
  address: Address;
  symbol: string;
  decimals: number;
}

export interface OracleObservation {
  sourceId: string;
  sourceKind: OracleKind;

  /**
   * Oracle answer before normalization.
   *
   * Example:
   * 1 USDC = $1.00000000 from an 8-decimal feed
   * answer = 100_000_000
   */
  answer: bigint;

  /**
   * Number of decimal places used by `answer`.
   */
  answerDecimals: number;

  /**
   * Unix timestamp in seconds at which the source was updated.
   */
  updatedAt: bigint;

  /**
   * Chain block from which the observation was obtained.
   */
  observedAtBlock: bigint;

  /**
   * Optional source round identifier.
   */
  roundId?: bigint;

  /**
   * Optional completed-round identifier.
   */
  answeredInRound?: bigint;

  /**
   * Optional source confidence in basis points.
   * 10,000 = full configured confidence.
   */
  confidenceBps?: bigint;
}

export interface OracleSource {
  readonly id: string;
  readonly kind: OracleKind;

  /**
   * Returns the USD price for exactly one whole token.
   */
  readUsdPrice(
    token: TokenMetadata,
    context: PricingContext,
  ): Promise<OracleObservation>;
}

export interface PricingContext {
  chainId: number;
  currentBlock: bigint;
  currentTimestamp: bigint;
}

export interface TokenOraclePolicy {
  token: Address;

  /**
   * Maximum permitted age for any accepted observation.
   */
  maxAgeSeconds: bigint;

  /**
   * Maximum difference between the freshest observation block
   * and the current block.
   */
  maxBlockLag: bigint;

  /**
   * Minimum number of valid sources required.
   */
  minimumValidSources: number;

  /**
   * Maximum difference between accepted oracle observations.
   */
  maximumDeviationBps: bigint;

  /**
   * Minimum accepted source confidence.
   */
  minimumConfidenceBps: bigint;

  /**
   * Source identifiers in priority order.
   */
  sourceIds: string[];

  /**
   * Aggregation rule used after validation.
   */
  aggregation: "MEDIAN" | "CONSERVATIVE_LOW" | "CONSERVATIVE_HIGH";
}

export interface PriceResult {
  token: TokenMetadata;

  /**
   * USD price of one whole token, normalized to 18 decimals.
   */
  priceUsdX18: bigint;

  observedAtBlock: bigint;
  updatedAt: bigint;

  sourcesUsed: string[];

  /**
   * Maximum source-to-source deviation after validation.
   */
  sourceDeviationBps: bigint;

  /**
   * Minimum confidence among accepted sources.
   */
  confidenceBps: bigint;
}

export interface PricePair {
  base: TokenMetadata;
  quote: TokenMetadata;

  /**
   * Quote-token units per one base token, scaled to 1e18.
   *
   * Example:
   * WETH/USDC = 3,000 quote units per base unit
   * basePerQuoteX18 represents 3000e18.
   */
  quotePerBaseX18: bigint;

  baseUsdX18: bigint;
  quoteUsdX18: bigint;
}

export class PricingError extends Error {
  constructor(
    public readonly code:
      | "TOKEN_NOT_REGISTERED"
      | "POLICY_NOT_REGISTERED"
      | "SOURCE_NOT_REGISTERED"
      | "INVALID_TOKEN_DECIMALS"
      | "INVALID_ORACLE_DECIMALS"
      | "NON_POSITIVE_ORACLE_PRICE"
      | "STALE_ORACLE_PRICE"
      | "ORACLE_BLOCK_LAG"
      | "INCOMPLETE_ORACLE_ROUND"
      | "LOW_ORACLE_CONFIDENCE"
      | "INSUFFICIENT_VALID_SOURCES"
      | "ORACLE_DEVIATION_EXCEEDED"
      | "DIVISION_BY_ZERO"
      | "NEGATIVE_VALUE"
      | "UNSAFE_DECIMAL_EXPONENT",
    message: string,
  ) {
    super(`${code}: ${message}`);
    this.name = "PricingError";
  }
}

export class PrecisionPricingEngine {
  private readonly tokens = new Map<string, TokenMetadata>();
  private readonly policies = new Map<string, TokenOraclePolicy>();
  private readonly sources = new Map<string, OracleSource>();

  constructor(
    tokens: TokenMetadata[],
    policies: TokenOraclePolicy[],
    sources: OracleSource[],
  ) {
    for (const token of tokens) {
      this.validateDecimals(token.decimals, "token");
      this.tokens.set(this.key(token.chainId, token.address), token);
    }

    for (const policy of policies) {
      this.policies.set(policy.token.toLowerCase(), policy);
    }

    for (const source of sources) {
      this.sources.set(source.id, source);
    }
  }

  getToken(chainId: number, address: Address): TokenMetadata {
    const token = this.tokens.get(this.key(chainId, address));

    if (!token) {
      throw new PricingError(
        "TOKEN_NOT_REGISTERED",
        `No metadata exists for ${chainId}:${address}`,
      );
    }

    return token;
  }

  /**
   * Resolves one token's USD price through the configured multi-oracle policy.
   */
  async getUsdPrice(
    tokenAddress: Address,
    context: PricingContext,
  ): Promise<PriceResult> {
    const token = this.getToken(context.chainId, tokenAddress);
    const policy = this.policies.get(tokenAddress.toLowerCase());

    if (!policy) {
      throw new PricingError(
        "POLICY_NOT_REGISTERED",
        `No oracle policy exists for ${token.symbol}`,
      );
    }

    const observations: OracleObservation[] = [];
    const failures: string[] = [];

    for (const sourceId of policy.sourceIds) {
      const source = this.sources.get(sourceId);

      if (!source) {
        failures.push(`${sourceId}:SOURCE_NOT_REGISTERED`);
        continue;
      }

      try {
        const observation = await source.readUsdPrice(token, context);
        this.validateObservation(observation, policy, context);
        observations.push(observation);
      } catch (error) {
        failures.push(
          `${sourceId}:${
            error instanceof Error ? error.message : String(error)
          }`,
        );
      }
    }

    if (observations.length < policy.minimumValidSources) {
      throw new PricingError(
        "INSUFFICIENT_VALID_SOURCES",
        [
          `${token.symbol} requires ${policy.minimumValidSources}`,
          `valid sources but received ${observations.length}.`,
          `Failures: ${failures.join(" | ")}`,
        ].join(" "),
      );
    }

    const normalized = observations.map((observation) => ({
      observation,
      priceUsdX18: scaleDecimals(
        observation.answer,
        observation.answerDecimals,
        PRICE_DECIMALS,
      ),
    }));

    const minimum = minBigInt(normalized.map((x) => x.priceUsdX18));
    const maximum = maxBigInt(normalized.map((x) => x.priceUsdX18));

    const deviationBps = calculateDeviationBps(minimum, maximum);

    if (deviationBps > policy.maximumDeviationBps) {
      throw new PricingError(
        "ORACLE_DEVIATION_EXCEEDED",
        `${token.symbol} source deviation ${deviationBps} bps exceeds ${
          policy.maximumDeviationBps
        } bps`,
      );
    }

    const selectedPrice = aggregatePrices(
      normalized.map((x) => x.priceUsdX18),
      policy.aggregation,
    );

    return {
      token,
      priceUsdX18: selectedPrice,
      observedAtBlock: minBigInt(
        observations.map((x) => x.observedAtBlock),
      ),
      updatedAt: minBigInt(observations.map((x) => x.updatedAt)),
      sourcesUsed: observations.map((x) => x.sourceId),
      sourceDeviationBps: deviationBps,
      confidenceBps: minBigInt(
        observations.map((x) => x.confidenceBps ?? BPS_DENOMINATOR),
      ),
    };
  }

  /**
   * Converts raw ERC-20 atomic units to USD fixed-point units.
   *
   * Result uses 18 USD decimals.
   */
  tokenAtomicToUsdX18(
    amountAtomic: bigint,
    token: TokenMetadata,
    priceUsdX18: bigint,
    rounding: Rounding = "DOWN",
  ): bigint {
    assertNonNegative(amountAtomic);
    assertPositive(priceUsdX18);

    const tokenScale = pow10(token.decimals);

    return mulDiv(
      amountAtomic,
      priceUsdX18,
      tokenScale,
      rounding,
    );
  }

  /**
   * Converts USD fixed-point value back into token atomic units.
   */
  usdX18ToTokenAtomic(
    usdValueX18: bigint,
    token: TokenMetadata,
    priceUsdX18: bigint,
    rounding: Rounding = "UP",
  ): bigint {
    assertNonNegative(usdValueX18);
    assertPositive(priceUsdX18);

    return mulDiv(
      usdValueX18,
      pow10(token.decimals),
      priceUsdX18,
      rounding,
    );
  }

  /**
   * Calculates the cross price:
   *
   * quote-token units required for one base token.
   */
  derivePairPrice(
    base: PriceResult,
    quote: PriceResult,
    rounding: Rounding = "DOWN",
  ): PricePair {
    assertPositive(quote.priceUsdX18);

    const quotePerBaseX18 = mulDiv(
      base.priceUsdX18,
      PRICE_SCALE,
      quote.priceUsdX18,
      rounding,
    );

    return {
      base: base.token,
      quote: quote.token,
      quotePerBaseX18,
      baseUsdX18: base.priceUsdX18,
      quoteUsdX18: quote.priceUsdX18,
    };
  }

  /**
   * Converts costs denominated in one token into another token.
   *
   * Used for:
   * gas token → flashloan token
   * relay token → flashloan token
   * profit token → USD
   */
  convertTokenAtomic(
    amountInAtomic: bigint,
    tokenIn: TokenMetadata,
    tokenInUsdX18: bigint,
    tokenOut: TokenMetadata,
    tokenOutUsdX18: bigint,
    rounding: Rounding = "UP",
  ): bigint {
    const usdValue = this.tokenAtomicToUsdX18(
      amountInAtomic,
      tokenIn,
      tokenInUsdX18,
      rounding,
    );

    return this.usdX18ToTokenAtomic(
      usdValue,
      tokenOut,
      tokenOutUsdX18,
      rounding,
    );
  }

  /**
   * Produces an executable venue price from actual leg amounts.
   *
   * amountIn / amountOut gives input-token units paid
   * per output-token unit received.
   */
  executablePriceX18(
    amountInAtomic: bigint,
    tokenIn: TokenMetadata,
    amountOutAtomic: bigint,
    tokenOut: TokenMetadata,
    rounding: Rounding = "UP",
  ): bigint {
    assertPositive(amountInAtomic);
    assertPositive(amountOutAtomic);

    const numerator =
      amountInAtomic *
      pow10(tokenOut.decimals) *
      PRICE_SCALE;

    const denominator =
      amountOutAtomic *
      pow10(tokenIn.decimals);

    return divide(numerator, denominator, rounding);
  }

  private validateObservation(
    observation: OracleObservation,
    policy: TokenOraclePolicy,
    context: PricingContext,
  ): void {
    this.validateDecimals(observation.answerDecimals, "oracle");

    if (observation.answer <= 0n) {
      throw new PricingError(
        "NON_POSITIVE_ORACLE_PRICE",
        `${observation.sourceId} returned ${observation.answer}`,
      );
    }

    if (
      observation.updatedAt <= 0n ||
      context.currentTimestamp < observation.updatedAt ||
      context.currentTimestamp - observation.updatedAt >
        policy.maxAgeSeconds
    ) {
      throw new PricingError(
        "STALE_ORACLE_PRICE",
        `${observation.sourceId} failed timestamp validation`,
      );
    }

    if (
      context.currentBlock < observation.observedAtBlock ||
      context.currentBlock - observation.observedAtBlock >
        policy.maxBlockLag
    ) {
      throw new PricingError(
        "ORACLE_BLOCK_LAG",
        `${observation.sourceId} exceeded block-lag policy`,
      );
    }

    if (
      observation.roundId !== undefined &&
      observation.answeredInRound !== undefined &&
      observation.answeredInRound < observation.roundId
    ) {
      throw new PricingError(
        "INCOMPLETE_ORACLE_ROUND",
        `${observation.sourceId} returned an incomplete round`,
      );
    }

    const confidence =
      observation.confidenceBps ?? BPS_DENOMINATOR;

    if (confidence < policy.minimumConfidenceBps) {
      throw new PricingError(
        "LOW_ORACLE_CONFIDENCE",
        `${observation.sourceId} confidence ${confidence} bps is below ${
          policy.minimumConfidenceBps
        } bps`,
      );
    }
  }

  private validateDecimals(
    decimals: number,
    type: "token" | "oracle",
  ): void {
    if (!Number.isInteger(decimals) || decimals < 0 || decimals > 36) {
      throw new PricingError(
        type === "token"
          ? "INVALID_TOKEN_DECIMALS"
          : "INVALID_ORACLE_DECIMALS",
        `${type} decimals ${decimals} are unsupported`,
      );
    }
  }

  private key(chainId: number, address: Address): string {
    return `${chainId}:${address.toLowerCase()}`;
  }
}

export type Rounding = "DOWN" | "UP" | "NEAREST";

/**
 * Scales an integer from one decimal precision to another.
 */
export function scaleDecimals(
  value: bigint,
  fromDecimals: number,
  toDecimals: number,
  rounding: Rounding = "DOWN",
): bigint {
  assertNonNegative(value);

  if (fromDecimals === toDecimals) {
    return value;
  }

  if (fromDecimals < toDecimals) {
    return value * pow10(toDecimals - fromDecimals);
  }

  return divide(
    value,
    pow10(fromDecimals - toDecimals),
    rounding,
  );
}

/**
 * Full-precision multiply-then-divide helper.
 *
 * JavaScript bigint does not overflow at uint256 boundaries, but downstream
 * Solidity implementations must use a full-precision mulDiv implementation.
 */
export function mulDiv(
  x: bigint,
  y: bigint,
  denominator: bigint,
  rounding: Rounding = "DOWN",
): bigint {
  if (denominator === 0n) {
    throw new PricingError(
      "DIVISION_BY_ZERO",
      "mulDiv denominator is zero",
    );
  }

  return divide(x * y, denominator, rounding);
}

export function divide(
  numerator: bigint,
  denominator: bigint,
  rounding: Rounding,
): bigint {
  if (denominator === 0n) {
    throw new PricingError(
      "DIVISION_BY_ZERO",
      "Division denominator is zero",
    );
  }

  const quotient = numerator / denominator;
  const remainder = numerator % denominator;

  if (remainder === 0n || rounding === "DOWN") {
    return quotient;
  }

  if (rounding === "UP") {
    return quotient + 1n;
  }

  return remainder * 2n >= denominator
    ? quotient + 1n
    : quotient;
}

export function calculateDeviationBps(
  minimum: bigint,
  maximum: bigint,
): bigint {
  assertPositive(minimum);

  return mulDiv(
    maximum - minimum,
    BPS_DENOMINATOR,
    minimum,
    "UP",
  );
}

export function aggregatePrices(
  prices: bigint[],
  mode: TokenOraclePolicy["aggregation"],
): bigint {
  if (prices.length === 0) {
    throw new PricingError(
      "INSUFFICIENT_VALID_SOURCES",
      "No prices were supplied for aggregation",
    );
  }

  const sorted = [...prices].sort((a, b) =>
    a < b ? -1 : a > b ? 1 : 0,
  );

  if (mode === "CONSERVATIVE_LOW") {
    return sorted[0];
  }

  if (mode === "CONSERVATIVE_HIGH") {
    return sorted[sorted.length - 1];
  }

  const middle = Math.floor(sorted.length / 2);

  if (sorted.length % 2 === 1) {
    return sorted[middle];
  }

  return (sorted[middle - 1] + sorted[middle]) / 2n;
}

export function pow10(decimals: number): bigint {
  if (!Number.isInteger(decimals) || decimals < 0 || decimals > 36) {
    throw new PricingError(
      "UNSAFE_DECIMAL_EXPONENT",
      `Unsupported decimal exponent ${decimals}`,
    );
  }

  return 10n ** BigInt(decimals);
}

function assertPositive(value: bigint): void {
  if (value <= 0n) {
    throw new PricingError(
      "NON_POSITIVE_ORACLE_PRICE",
      `Expected positive value; received ${value}`,
    );
  }
}

function assertNonNegative(value: bigint): void {
  if (value < 0n) {
    throw new PricingError(
      "NEGATIVE_VALUE",
      `Expected non-negative value; received ${value}`,
    );
  }
}

function minBigInt(values: bigint[]): bigint {
  if (values.length === 0) {
    throw new PricingError(
      "INSUFFICIENT_VALID_SOURCES",
      "Cannot determine minimum of an empty collection",
    );
  }

  return values.reduce((minimum, value) =>
    value < minimum ? value : minimum,
  );
}

function maxBigInt(values: bigint[]): bigint {
  if (values.length === 0) {
    throw new PricingError(
      "INSUFFICIENT_VALID_SOURCES",
      "Cannot determine maximum of an empty collection",
    );
  }

  return values.reduce((maximum, value) =>
    value > maximum ? value : maximum,
  );
}

File: omega_v5/pricing/precision_pricing.py
#!/usr/bin/env python3
# ==============================================================================
# precision_pricing.py — Python port of APEX Ω Precision Pricing Engine
#
# Makes the canonical TS logic (BigInt-only, 18-decimal scale, strict oracle
# validation, mulDiv, aggregation) fully compatible with the Omega V5 pipeline.
#
# All monetary math uses native Python int (arbitrary precision, no float/Decimal
# for core price paths). Matches the TS contract exactly for cross-language
# consistency in the execution pipeline.
#
# Responsibilities (from TS):
# 1. Resolve token decimals.
# 2. Resolve oracle decimals.
# 3. Normalize every oracle price into PRICE_SCALE (1e18).
# 4. Convert token atomic units into USD fixed-point units.
# 5. Convert USD values back into token atomic units.
# 6. Derive TOKEN_A / TOKEN_B cross prices.
# 7. Reject stale, invalid, incomplete, or divergent prices.
# 8. Preserve integer precision throughout the execution pipeline.
# ==============================================================================

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Protocol, Tuple

# ── Constants (exact match to TS) ─────────────────────────────────────────────
PRICE_DECIMALS: int = 18
PRICE_SCALE: int = 10 ** PRICE_DECIMALS

BPS_DENOMINATOR: int = 10_000


class OracleKind(str, Enum):
    CHAINLINK = "CHAINLINK"
    DEX_TWAP = "DEX_TWAP"
    PROTOCOL_RATE = "PROTOCOL_RATE"
    MANUAL_DISABLED = "MANUAL_DISABLED"


@dataclass
class TokenMetadata:
    chain_id: int
    address: str  # 0x...
    symbol: str
    decimals: int


@dataclass
class OracleObservation:
    source_id: str
    source_kind: OracleKind

    # Oracle answer before normalization (e.g. 8-decimal feed -> 100000000 for $1)
    answer: int
    answer_decimals: int

    updated_at: int  # unix seconds
    observed_at_block: int

    round_id: Optional[int] = None
    answered_in_round: Optional[int] = None
    confidence_bps: Optional[int] = None


class OracleSource(Protocol):
    """Returns the USD price for exactly one whole token."""
    id: str
    kind: OracleKind

    def read_usd_price(
        self, token: TokenMetadata, context: PricingContext
    ) -> OracleObservation:
        ...


@dataclass
class PricingContext:
    chain_id: int
    current_block: int
    current_timestamp: int


@dataclass
class TokenOraclePolicy:
    token: str  # address lower
    max_age_seconds: int
    max_block_lag: int
    minimum_valid_sources: int
    maximum_deviation_bps: int
    minimum_confidence_bps: int
    source_ids: List[str]
    aggregation: str  # "MEDIAN" | "CONSERVATIVE_LOW" | "CONSERVATIVE_HIGH"


@dataclass
class PriceResult:
    token: TokenMetadata
    price_usd_x18: int  # USD price of 1 whole token @ 18 decimals
    observed_at_block: int
    updated_at: int
    sources_used: List[str]
    source_deviation_bps: int
    confidence_bps: int


@dataclass
class PricePair:
    base: TokenMetadata
    quote: TokenMetadata
    quote_per_base_x18: int  # quote units per 1 base, scaled 1e18
    base_usd_x18: int
    quote_usd_x18: int


class PricingError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(f"{code}: {message}")
        self.code = code
        self.name = "PricingError"


class Rounding(str, Enum):
    DOWN = "DOWN"
    UP = "UP"
    NEAREST = "NEAREST"


class PrecisionPricingEngine:
    """Python port of the TS PrecisionPricingEngine.

    Registers tokens, policies, and sources. Provides exact integer pricing
    for the arbitrage/execution pipeline.
    """

    def __init__(
        self,
        tokens: List[TokenMetadata],
        policies: List[TokenOraclePolicy],
        sources: List[OracleSource],
    ):
        self.tokens: Dict[str, TokenMetadata] = {}
        self.policies: Dict[str, TokenOraclePolicy] = {}
        self.sources: Dict[str, OracleSource] = {}

        for token in tokens:
            self._validate_decimals(token.decimals, "token")
            key = self._key(token.chain_id, token.address)
            self.tokens[key] = token

        for policy in policies:
            self.policies[policy.token.lower()] = policy

        for source in sources:
            self.sources[source.id] = source

    def get_token(self, chain_id: int, address: str) -> TokenMetadata:
        key = self._key(chain_id, address)
        token = self.tokens.get(key)
        if not token:
            raise PricingError(
                "TOKEN_NOT_REGISTERED",
                f"No metadata exists for {chain_id}:{address}",
            )
        return token

    def get_usd_price(
        self, token_address: str, context: PricingContext
    ) -> PriceResult:
        token = self.get_token(context.chain_id, token_address)
        policy = self.policies.get(token_address.lower())

        if not policy:
            raise PricingError(
                "POLICY_NOT_REGISTERED",
                f"No oracle policy exists for {token.symbol}",
            )

        observations: List[OracleObservation] = []
        failures: List[str] = []

        for source_id in policy.source_ids:
            source = self.sources.get(source_id)
            if not source:
                failures.append(f"{source_id}:SOURCE_NOT_REGISTERED")
                continue

            try:
                observation = source.read_usd_price(token, context)
                self._validate_observation(observation, policy, context)
                observations.append(observation)
            except Exception as exc:  # broad to capture PricingError + source errors
                msg = str(exc)
                failures.append(f"{source_id}:{msg}")

        if len(observations) < policy.minimum_valid_sources:
            raise PricingError(
                "INSUFFICIENT_VALID_SOURCES",
                f"{token.symbol} requires {policy.minimum_valid_sources} "
                f"valid sources but received {len(observations)}. "
                f"Failures: {' | '.join(failures)}",
            )

        normalized = [
            {
                "observation": obs,
                "price_usd_x18": scale_decimals(
                    obs.answer, obs.answer_decimals, PRICE_DECIMALS
                ),
            }
            for obs in observations
        ]

        prices = [n["price_usd_x18"] for n in normalized]
        minimum = min_bigint(prices)
        maximum = max_bigint(prices)

        deviation_bps = calculate_deviation_bps(minimum, maximum)

        if deviation_bps > policy.maximum_deviation_bps:
            raise PricingError(
                "ORACLE_DEVIATION_EXCEEDED",
                f"{token.symbol} source deviation {deviation_bps} bps exceeds "
                f"{policy.maximum_deviation_bps} bps",
            )

        selected_price = aggregate_prices(
            prices, policy.aggregation  # type: ignore[arg-type]
        )

        return PriceResult(
            token=token,
            price_usd_x18=selected_price,
            observed_at_block=min_bigint([o.observed_at_block for o in observations]),
            updated_at=min_bigint([o.updated_at for o in observations]),
            sources_used=[o.source_id for o in observations],
            source_deviation_bps=deviation_bps,
            confidence_bps=min_bigint(
                [o.confidence_bps or BPS_DENOMINATOR for o in observations]
            ),
        )

    def token_atomic_to_usd_x18(
        self,
        amount_atomic: int,
        token: TokenMetadata,
        price_usd_x18: int,
        rounding: Rounding = Rounding.DOWN,
    ) -> int:
        assert_non_negative(amount_atomic)
        assert_positive(price_usd_x18)

        token_scale = pow10(token.decimals)
        return mul_div(amount_atomic, price_usd_x18, token_scale, rounding)

    def usd_x18_to_token_atomic(
        self,
        usd_value_x18: int,
        token: TokenMetadata,
        price_usd_x18: int,
        rounding: Rounding = Rounding.UP,
    ) -> int:
        assert_non_negative(usd_value_x18)
        assert_positive(price_usd_x18)

        return mul_div(
            usd_value_x18, pow10(token.decimals), price_usd_x18, rounding
        )

    def derive_pair_price(
        self,
        base: PriceResult,
        quote: PriceResult,
        rounding: Rounding = Rounding.DOWN,
    ) -> PricePair:
        assert_positive(quote.price_usd_x18)

        quote_per_base_x18 = mul_div(
            base.price_usd_x18, PRICE_SCALE, quote.price_usd_x18, rounding
        )

        return PricePair(
            base=base.token,
            quote=quote.token,
            quote_per_base_x18=quote_per_base_x18,
            base_usd_x18=base.price_usd_x18,
            quote_usd_x18=quote.price_usd_x18,
        )

    def convert_token_atomic(
        self,
        amount_in_atomic: int,
        token_in: TokenMetadata,
        token_in_usd_x18: int,
        token_out: TokenMetadata,
        token_out_usd_x18: int,
        rounding: Rounding = Rounding.UP,
    ) -> int:
        usd_value = self.token_atomic_to_usd_x18(
            amount_in_atomic, token_in, token_in_usd_x18, rounding
        )
        return self.usd_x18_to_token_atomic(
            usd_value, token_out, token_out_usd_x18, rounding
        )

    def executable_price_x18(
        self,
        amount_in_atomic: int,
        token_in: TokenMetadata,
        amount_out_atomic: int,
        token_out: TokenMetadata,
        rounding: Rounding = Rounding.UP,
    ) -> int:
        assert_positive(amount_in_atomic)
        assert_positive(amount_out_atomic)

        numerator = amount_in_atomic * pow10(token_out.decimals) * PRICE_SCALE
        denominator = amount_out_atomic * pow10(token_in.decimals)
        return divide(numerator, denominator, rounding)

    # ── Private validation (exact TS logic) ───────────────────────────────────
    def _validate_observation(
        self,
        observation: OracleObservation,
        policy: TokenOraclePolicy,
        context: PricingContext,
    ) -> None:
        self._validate_decimals(observation.answer_decimals, "oracle")

        if observation.answer <= 0:
            raise PricingError(
                "NON_POSITIVE_ORACLE_PRICE",
                f"{observation.source_id} returned {observation.answer}",
            )

        if (
            observation.updated_at <= 0
            or context.current_timestamp < observation.updated_at
            or context.current_timestamp - observation.updated_at
            > policy.max_age_seconds
        ):
            raise PricingError(
                "STALE_ORACLE_PRICE",
                f"{observation.source_id} failed timestamp validation",
            )

        if (
            context.current_block < observation.observed_at_block
            or context.current_block - observation.observed_at_block
            > policy.max_block_lag
        ):
            raise PricingError(
                "ORACLE_BLOCK_LAG",
                f"{observation.source_id} exceeded block-lag policy",
            )

        if (
            observation.round_id is not None
            and observation.answered_in_round is not None
            and observation.answered_in_round < observation.round_id
        ):
            raise PricingError(
                "INCOMPLETE_ORACLE_ROUND",
                f"{observation.source_id} returned an incomplete round",
            )

        confidence = observation.confidence_bps or BPS_DENOMINATOR
        if confidence < policy.minimum_confidence_bps:
            raise PricingError(
                "LOW_ORACLE_CONFIDENCE",
                f"{observation.source_id} confidence {confidence} bps is below "
                f"{policy.minimum_confidence_bps} bps",
            )

    def _validate_decimals(self, decimals: int, type_: str) -> None:
        if not isinstance(decimals, int) or decimals < 0 or decimals > 36:
            code = (
                "INVALID_TOKEN_DECIMALS"
                if type_ == "token"
                else "INVALID_ORACLE_DECIMALS"
            )
            raise PricingError(
                code, f"{type_} decimals {decimals} are unsupported"
            )

    def _key(self, chain_id: int, address: str) -> str:
        return f"{chain_id}:{address.lower()}"


# ── Pure integer math helpers (exact port of TS) ──────────────────────────────
def scale_decimals(
    value: int,
    from_decimals: int,
    to_decimals: int,
    rounding: Rounding = Rounding.DOWN,
) -> int:
    assert_non_negative(value)

    if from_decimals == to_decimals:
        return value

    if from_decimals < to_decimals:
        return value * pow10(to_decimals - from_decimals)

    return divide(value, pow10(from_decimals - to_decimals), rounding)


def mul_div(
    x: int, y: int, denominator: int, rounding: Rounding = Rounding.DOWN
) -> int:
    if denominator == 0:
        raise PricingError("DIVISION_BY_ZERO", "mulDiv denominator is zero")
    return divide(x * y, denominator, rounding)


def divide(numerator: int, denominator: int, rounding: Rounding) -> int:
    if denominator == 0:
        raise PricingError("DIVISION_BY_ZERO", "Division denominator is zero")

    quotient = numerator // denominator
    remainder = numerator % denominator

    if remainder == 0 or rounding == Rounding.DOWN:
        return quotient

    if rounding == Rounding.UP:
        return quotient + 1

    # NEAREST
    return quotient + 1 if remainder * 2 >= denominator else quotient


def calculate_deviation_bps(minimum: int, maximum: int) -> int:
    assert_positive(minimum)
    return mul_div(maximum - minimum, BPS_DENOMINATOR, minimum, Rounding.UP)


def aggregate_prices(
    prices: List[int], mode: str
) -> int:
    if not prices:
        raise PricingError(
            "INSUFFICIENT_VALID_SOURCES", "No prices were supplied for aggregation"
        )

    sorted_prices = sorted(prices)

    if mode == "CONSERVATIVE_LOW":
        return sorted_prices[0]
    if mode == "CONSERVATIVE_HIGH":
        return sorted_prices[-1]

    # MEDIAN
    middle = len(sorted_prices) // 2
    if len(sorted_prices) % 2 == 1:
        return sorted_prices[middle]
    return (sorted_prices[middle - 1] + sorted_prices[middle]) // 2


def pow10(decimals: int) -> int:
    if not isinstance(decimals, int) or decimals < 0 or decimals > 36:
        raise PricingError(
            "UNSAFE_DECIMAL_EXPONENT", f"Unsupported decimal exponent {decimals}"
        )
    return 10 ** decimals


def assert_positive(value: int) -> None:
    if value <= 0:
        raise PricingError(
            "NON_POSITIVE_ORACLE_PRICE", f"Expected positive value; received {value}"
        )


def assert_non_negative(value: int) -> None:
    if value < 0:
        raise PricingError(
            "NEGATIVE_VALUE", f"Expected non-negative value; received {value}"
        )


def min_bigint(values: List[int]) -> int:
    if not values:
        raise PricingError(
            "INSUFFICIENT_VALID_SOURCES",
            "Cannot determine minimum of an empty collection",
        )
    return min(values)


def max_bigint(values: List[int]) -> int:
    if not values:
        raise PricingError(
            "INSUFFICIENT_VALID_SOURCES",
            "Cannot determine maximum of an empty collection",
        )
    return max(values)


# ── Convenience factory for pipeline integration ──────────────────────────────
def create_default_engine(
    tokens: List[TokenMetadata],
    policies: List[TokenOraclePolicy],
    sources: List[OracleSource],
) -> PrecisionPricingEngine:
    """Factory used by the pipeline to wire the precision engine."""
    return PrecisionPricingEngine(tokens, policies, sources)
